src/test/scala/chiselTests/properties/PropertySpec.scala

```diff
@@ -617,4 +617,44 @@ class PropertySpec extends ChiselFlatSpec with MatchesAndOmits {
       lit.isLit shouldBe true
     })
   }
+
+  behavior.of("PropertyArithmeticOps")
+
+  it should "support expressions in temporaries, wires, and ports" in {
+    val chirrtl = ChiselStage.emitCHIRRTL(new RawModule {
+      val a = IO(Input(Property[Int]()))
+      val b = IO(Input(Property[Int]()))
+      val c = IO(Output(Property[Int]()))
+      val d = IO(Output(Property[Int]()))
+      val e = IO(Output(Property[Int]()))
+
+      val t = a + b
+
+      val w = WireInit(t)
+
+      c := t
+      d := t + a
+      e := w + (a + b)
+    })
+
+    matchesAndOmits(chirrtl)(
+      "propassign w, integer_add(a, b)",
+      "propassign c, integer_add(a, b)",
+      "propassign d, integer_add(integer_add(a, b), a)",
+      "propassign e, integer_add(w, integer_add(a, b))"
+    )()
```

Comment:
This is interesting as it demonstrates an emission strategy for more compact FIRRTL. I.e., we could emit this way instead of with many, many nodes and the FIRRTL would be conceptually much closer to the underlying Chisel expressions.

Reply:
Yeah, I think this is the main thing I'm interested in getting feedback from in this draft PR. It's quite different from how expressions like this normally work, but I like this emission because a) we have always aimed for static single assignment of properties, b) it is completely trivial to parse in the FIRRTL compiler, and standard CSE makes short work of any duplicated subexpressions.

To expand a bit on how this is implemented... here's roughly the way it works for other Data:

Create new Data subtype object in memory
-> give it an OpBinding 
-> push a DefPrim command in the Builder
-> when the Data is used as an rvalue, make a Node

Compared to the new approach I tried out here:

Create a new Property object in memory
-> give it a PropExprBinding that holds the primitive operation and the operands
-> *dont* push any commands in the Builder
-> when the Data is used as an rvalue, make a PropExpr

The new PropExpr Arg is handled differently in the Converter and Serializer. Rather than the Converter seeing a DefPrim, converting it, and defining a new Node, the Converter essentially passes through the same information that was present in the PropExprBinding and PropExpr. The Serializer then recursively serializes the expression tree in whatever rvalue context it is used (only the RHS of propassign or as a sub-expression, presently).

I like the result, but since it diverges from how we normally use Bindings and Args, I am curious if this is a valid implementation or if I should look into other techniques.

Reply:
we have no plan for any sort of last-connect semantics for properties, right? is all the complexity in the existing system described really there to enable that?

Reply:
> we have no plan for any sort of last-connect semantics for properties, right?

Yep, see 5. and 6. here: https://github.com/chipsalliance/firrtl-spec/blob/main/spec.md#property-assignments

> is all the complexity in the existing system described really there to enable that?

The use of Node for pretty much any rvalue pre-dates me diving into Chisel, so I'm not sure the historical reasons this has been done. I think making Nodes everywhere certainly makes some things simpler. I don't know if I'd say the existing approach exists to enable last connect though, since that's more about what you can do with lvalues than how you represent rvalues internally.

Reply:
> This is interesting as it demonstrates an emission strategy for more compact FIRRTL. I.e., we could emit this way instead of with many, many nodes and the FIRRTL would be conceptually much closer to the underlying Chisel expressions.

> Yeah, I think this is the main thing I'm interested in getting feedback from in this draft PR. It's quite different from how expressions like this normally work, but I like this emission because a) we have always aimed for static single assignment of properties, b) it is completely trivial to parse in the FIRRTL compiler, and standard CSE makes short work of any duplicated subexpressions.

I'm all for more compact FIRRTL that is closer to the original Chisel, but there are some important considerations:
1. What about when the user _wants_ to name the subexpression? Synthetic names are terrible so if the user assigns some subexpression to a val, we want that name which implies a `node` or something very much like a node.
2. We also need to think about how to track different source locators. `e := w + (a + b)` has 3 source locators, at least 3 columns, but if you had function calls they could different in line and source file as well. How do we record that information? It's possible with alternative source locator encoding, eg. SourceMaps, but it's generally important information.

Reply:
Thanks Jack, interesting points.

1. I know that this has been the Chisel -> Verilog contract, but I'm wondering if we don't need to uphold that contract any more. For one, using `val` names to control what is visible in the Verilog output has been a major source of frustration, both for users and FIRRTL compiler developers. For properties, we already have a different contract--the names of your classes and their output ports are what is guaranteed to be visible in the output (by the implementation, not the spec!). Maybe that is actually a decent contract, and we can treat intermediate property expressions more like how software compilers treat them--they have a name in the source, and you can look up the associated value while debugging if necessary (which gets into the second point).
2. I need to look into how Chisel is doing source locations, but there is probably something we can do here. All the source locations are there in memory, so if there is a format to emit compound source locations into, we can use it.

In general--I've been interested in this approach because the property stuff is new, and I'm trying to see if we can use an implementation that is closer to what we actually want for both Chisel and the FIRRTL compiler. The compiler is already operating on SSA form internally, so I think the value of nodes in the textual FIRRTL is pretty low. In order for this to work for end users, we need to figure out how to map the source locations in-memory in Chisel to the appropriate SSA values after parsing. Otherwise, I think this approach doesn't lose much for properties, and more closely matches how the compiler wants to see them.

On the flip side, we can easily go the more traditional Chisel route of spilling all intermediate expressions into nodes/wires if we decide that is preferable. I just wanted to try out something new and share it.

Reply:
> we can easily go the more traditional Chisel route of spilling all intermediate expressions into nodes/wires if we decide that is preferable

Discussed this synchronously today, and seems the consensus is this is preferable, so I will amend the first commit of this PR.

This did kick off an interesting discussion that we should continue, which might lead to a better FIRRTL spec around (sub-)expressions.

Reply:
One more nail in the coffin... the current approach doesn't support BoringUtils.bore from a property expression, which is something this should definitely support. I'll add a test for that as I update the branch.